DOCS-136 update convex guide (#667)
clerk/clerk-docs: docs/integrations/databases/convex.mdx

description: Learn how to integrate Clerk into your Convex application.
---

<TutorialHero
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "https://clerk.com/docs/quickstarts/setup-clerk",
    },
    {
      title: "Create a React + Convex application",
      link: "https://docs.convex.dev/quickstart/react",
    },
  ]}
>
- Create a JWT Template based on Convex
- Configure Convex with the Clerk issuer domain
- Install and configure Clerk's React SDK
- Configure the Clerk and Convex providers
</TutorialHero>

# Integrate Convex with Clerk

> Convex is a backend-as-a-service (BAAS) platform that allows you to build applications without having to worry about the backend. Convex provides a GraphQL API, authentication, and a database. Paired with Clerk's user authentication and management features, you can build a powerful application with minimal effort.

Reviewer: Hey @alexisintech, I'm really curious, how did you come up with Convex providing a "GraphQL API"?
```suggestion
Convex is the full-stack TypeScript development platform. With Convex you get to build a backend with a provided realtime database, file storage, text search, scheduling and more. Paired with Clerk's user authentication and management features, you can build a powerful application with minimal effort.
```

Author: Hey @xixixao this sentence was actually given to me by GitHub Copilot (embarrassed) but this is exactly why I wanted your eyes on it! I've never worked with Convex, so I truly apologize for this mishap. I'll commit your suggestion now!

  alt="The 'Create new template' page of the JWT Templates page in the Clerk Dashboard. There is a red box surrounding the 'Issuer' section."
/>

In your `convex` folder, add an `auth.config.js` file with the following configuration:

> ```ts title="convex/auth.config.js"
> export default {
>   providers: [
>     {
>       domain: "https://your-issuer-url.clerk.accounts.dev/",
>       applicationID: "convex",
>     },
>   ]
> };
> ```

Reviewer: This isn't rendering correctly, probably because `title` instead of `filename`
```suggestion
```ts filename="convex/auth.config.js"
export default {
  providers: [
    {
      domain: "https://your-issuer-url.clerk.accounts.dev/",
      applicationID: "convex",
    },
  ]
};
```
```

Author: great catch!!

Replace the `domain` string with the **Issuer** URL you copied.

### Deploy your changes to Convex

Run `npx convex dev` to automatically sync your configuration to your backend.

### Install `@clerk/clerk-react`

Install Clerk's React SDK by running the following command in your terminal:

<CodeBlockTabs type="installer" options={["npm", "yarn",  "pnpm" ]}>
  ```bash filename="terminal"
  npm install @clerk/clerk-react
  ```
  ```bash filename="terminal"
  yarn add @clerk/clerk-react
  ```
  
  ```bash filename="terminal"
  pnpm add @clerk/clerk-react
  ```

</CodeBlockTabs>

### Set environment keys

> In your React project's root folder, you may have an `.env` file alongside `package.json` and other configuration files. If you don't see it, create it.
> 
> Add the following code to your `.env` file to set your public key.

Reviewer: Vite and Convex use `.env.local`, not `.env`
```suggestion
In your React project's root folder, you may have an `.env.local` file alongside `package.json` and other configuration files. If you don't see it, create it.

Add the following code to your `.env.local` file to set your public key.
```

**Pro tip!** If you are signed into your [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=api-keys), you can copy your publishable key below.

<InjectKeys>

> ```sh filename=".env"

Reviewer: ```suggestion
```sh filename=".env.local"
```

VITE_CLERK_PUBLISHABLE_KEY={{pub_key}}
```

</InjectKeys>

### Import the Publishable Key

You will need to import your Publishable Key into your application. You can add an `if` statement to check that it is imported and that it exists. This will prevent running the application without the Publishable Key, and will also prevent TypeScript errors.

```tsx filename="src/main.tsx" {6-7, 9-11}
import React from 'react'
import ReactDOM from 'react-dom/client'
import App from './App.tsx'
import './index.css'

// Import your publishable key
const PUBLISHABLE_KEY = import.meta.env.VITE_CLERK_PUBLISHABLE_KEY

if (!PUBLISHABLE_KEY) {
  throw new Error("Missing Publishable Key")
}

ReactDOM.createRoot(document.getElementById('root')!).render(
  <React.StrictMode>
      <App />
  </React.StrictMode>,

> )

Reviewer: I find having this whole step (from line 129) with code that doesn't match the onboarding flow, confusing. The code here is included in the next step. Can we remove it? We can leave the throwing (although I'm pretty sure ClerkProvider will throw anyway)

Author: Great point ! I copied and pasted this from the React quickstart in a haste, and you are completely right that this step is misplaced. I have removed it :) Great catch.

```

### Configure the Clerk and Convex providers

Both Clerk and Convex have Provider components that are required to wrap your React application to provide the authentication and client context.

You may have already had a `<ConvexProvider>` configured. Be sure that `<ClerkProvider>` wraps `ConvexProviderWithClerk`, and that `useAuth` is passed to `ConvexProviderWithClerk`.

> ```ts title="src/main.tsx"

Reviewer: ```suggestion
```ts filename="src/main.tsx"
```

import React from "react";
import ReactDOM from "react-dom/client";
import App from "./App";
import "./index.css";
import { ClerkProvider, useAuth } from "@clerk/clerk-react";
import { ConvexProviderWithClerk } from "convex/react-clerk";
import { ConvexReactClient } from "convex/react";

// Import your publishable key
const PUBLISHABLE_KEY = import.meta.env.VITE_CLERK_PUBLISHABLE_KEY

if (!PUBLISHABLE_KEY) {
  throw new Error("Missing Publishable Key")
}

const convex = new ConvexReactClient(import.meta.env.VITE_CONVEX_URL as string);

ReactDOM.createRoot(document.getElementById("root")!).render(
  <React.StrictMode>
    <ClerkProvider publishableKey={PUBLISHABLE_KEY}>
      <ConvexProviderWithClerk client={convex} useAuth={useAuth}>
        <App />
      </ConvexProviderWithClerk>
    </ClerkProvider>
  </React.StrictMode>
);
```

### Access user identity in Convex queries and mutations

You can access the user information from the JWT in Convex queries and mutations.
Use the `ctx.auth.getUserIdentity()` which returns the parsed information from the JWT, or `null` if the client isn't authenticated.

> ```ts title=convex/getUser.ts

Reviewer: ```suggestion
```ts filename=convex/getUser.ts
```

import type { UserIdentity } from 'convex/server';
import { query } from './_generated/server';

> export default query(async (ctx): Promise<string | null> => {

Reviewer: I did not notice this before but this return type annotation is not needed
```suggestion
export default query(async (ctx) => {
```
